### PR TITLE
[refactor] Create base BatchInstanceLoader, port BatchAssetRecordLoader

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This creates a new BatchInstanceLoader abstraction, and ports BatchAssetRecordLoader to it.

This opens the door for the next PR to add similar loaders for fetching runs and asset check records without having to duplicate a ton of code.

The longer story of why I want this is that the AssetGraphView class needs to be able to query for these sorts of things for all sorts of different reasons. For example, we need asset records to check the status of assets, but it's also used when resolving the status of a check.

So if we can have the AssetGraphView class have a few loader objects attached to it (asset record, runs, asset check records), we can have a shared caching layer for any use case that needs to take advantage of these things.

## Changelog [New | Bug | Docs]

NOCHANGELOG

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24237?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24237** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24237?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#24202** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24200** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24200?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24111** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24111?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23910** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23910?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23907** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23905** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23904** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23904?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23843** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23843?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#23842](https://github.com/dagster-io/dagster/pull/23842) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23842?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by schrockn — COMMENTED** [review]
@alangenfeld @gibsondan what is the status of using asyncio and the `DataLoader` in the code base? Want to make sure we are converging here.

File: python_modules/dagster/dagster/_core/asset_graph_view/batch_instance_loader.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        return key in self._cache
+
+    @abstractmethod
+    def _query(self, keys: Set[K]) -> Mapping[K, V]: ...
+
+
+class BatchRunLoader(BatchInstanceLoader[str, DagsterRun]):
+    def _query(self, keys: AbstractSet[str]) -> Mapping[str, DagsterRun]:
+        runs = self._instance.get_runs(filters=RunsFilter(run_ids=list(keys)))
+        return {run.run_id: run for run in runs}
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/asset_graph_view/batch_instance_loader.py:61]: this was replaced by a `DataLoader` approach here https://github.com/dagster-io/dagster/pull/21753

**Review by alangenfeld — COMMENTED** [review]
> what is the status of using asyncio and the DataLoader in the code base? Want to make sure we are converging here.

`InstanceLoadableBy` exists and is in use for a few cases across the codebases. This abstraction would sort of be going backwards on moving towards asyncio based batching. 

That said based on the summary it sounds like the target here are use cases not in GraphQL where we can easily make functions `async` and do event loop based batching. The question than is can we use `asyncio` in the target `AssetGraphView` callsites potentially via `asyncio.run` -ing batches of concurrent work.

**Comment by OwenKephart** [comment]
I didn't know the DataLoader abstraction existed when writing this -- I think it's definitely strictly better. Definitely think there are opportunities for taking advantage of async stuff in the AssetGraphView, at the very least for a sort of pre-fetching step where we go out and query batches of information across a few different types.

I'll update the PR accordingly

**Comment by schrockn** [comment]
Well I think we need to add synchronous variants to `InstanceLoadableBy` to make this usable in these contexts. @alangenfeld wdyt?

**Comment by alangenfeld** [comment]
my bias would be to try to introduce `asyncio` in to more places since i think that will evolve better but i am not against adding a sync api

**Comment by OwenKephart** [comment]
closed this in favor of: https://github.com/dagster-io/dagster/pull/24276

**Pull request closed without merging**